### PR TITLE
Extend and improve color contrasting

### PR DESCRIPTION
**Pull request by Meorawr**
This applies an absolute unit of changes to improve our color contrast setting. In terms of end-user changes, the following has been done:

- The characteristics page now applies the color contrast setting to the name text, class text, and eye color texts.
- The characteristics page shows a color swatch next to class and eye color fields that displays the unaltered color without contrast adjustments enforced.
- The color browser now displays a warning if the selected color is unlikely to be readable by other players against dark backgrounds, and offers an alternative suggestion instead.
- The color browser now automatically updates the selected color once a full 6 byte hex code has been input into the editbox, either by typing or pasting. The visuals of the field have been updated to match Blizzard's own added hex color input field too.
- The color browser preset menu now displays colors in a swatch on the buttons themselves, rather than coloring the text. This does not apply to Classic.
- For custom presets, the color browser no longer requires that you select a color first before renaming or deleting it. Instead these options are provided in a submenu of the preset entry.
- The personality traits have had an inconsistency fixed where the text color used by presets differed from the bar color. Both colors have been altered to a version that's closer to what the text was, and should be automatically readable by most players without contrast adjustments.
- Personality trait bars now display a thumb in read-mode to better separate the left and right sections. This is intended mostly to benefit users who may have trouble differentiating the chosen bar colors.

In terms of internal changes;

- The old text truncation template (TRP3_TruncatedTextTemplate) has been removed. The reason this template existed was because historically fontstrings couldn't display tooltips, but this hasn't been the case for two years. All usages of this template (of which none are in extended) have been migrated.
- The fontstring truncation mixin has been renamed to TRP3_TruncatedTextMixin, now that the name used by the old template is freed up.
- A new mixin has been added (TRP3_ReadableTextMixin) which adds a single SetReadableTextColor method to an object. This can be used to more easily apply color contrast adjustments for custom colors.
- A new color swatch template (TRP3_ColorSwatchTemplate) has been added for general use. This *should* display quite neatly at various UI scales and window sizes, but perfection is not guaranteed.
- The API around color contrast has been slightly cleaned up.
- The color browser was essentially rewritten from the ground up.

Piccies.

Color swatch and applied contrast for the "Eye color" field (hard to see, but it's there - trust).
Wow_2024-08-03_17-52-13

Tooltip when hovering a swatch that differs from the text color.
Wow_2024-08-03_17-52-17

Adjusted colors for preset traits to improve readability and be consistent.
Wow_2024-08-03_17-52-24

Readability warning in the color browser, along with updated hex input field.
Wow_2024-08-03_17-52-42

Nested submenus for managing custom presets, as well as the swatch instead of text coloring.
Wow_2024-08-03_17-52-49

### CODE REVIEW

File: totalRP3/Modules/Register/Characters/RegisterUICharacteristics.xml [hunk 1]
```diff
@@ -143,26 +134,34 @@ https://raw.githubusercontent.com/Meorawr/wow-ui-schema/main/UI.xsd">
 				</Texture>
 			</Layer>
 			<Layer level="BORDER" textureSubLevel="-1">
-				<Texture parentKey="OppositeFill" file="Interface\TargetingFrame\UI-StatusBar" hWrapMode="REPEAT">
+				<Texture parentKey="OppositeFill" atlas="nameplates-bar-fill" hWrapMode="REPEAT">
+				<!-- <Texture parentKey="OppositeFill" file="Interface\TargetingFrame\UI-StatusBar" hWrapMode="REPEAT"> -->
```
Comment (Meorawr) [inline totalRP3/Modules/Register/Characters/RegisterUICharacteristics.xml:138]: Note; need to go back and fix Classic as the named atlas doesn't exist.

**Comment by Raenore** [comment]
Woah are we in the year of 2024?

**Comment by Raenore** [comment]
Sorry for the wait.

I'll start testing this today and check for breakages.

**Review by Raenore — DISMISSED** [review]
Left trait icons seem to be broken on profiles.

Before:



After:


Edit mode does show them:

File: totalRP3/Modules/Register/Main/RegisterTemplates.lua [hunk 2]
```diff
@@ -10,3 +12,111 @@ end
...
+	self.LeftText:SetText(text);
+end
+
+function TRP3_RegisterTraitLineMixin:SetRightText(text)
+	text = string.gsub(text or "", "%s+", " ");
+	self.RightText:SetText(text);
+end
+
+function TRP3_RegisterTraitLineMixin:SetLeftIcon(icon)
+	self.RightIcon:SetIconTexture(icon);
```
Comment (Raenore) [inline totalRP3/Modules/Register/Main/RegisterTemplates.lua:37]:
```suggestion
	self.LeftIcon:SetIconTexture(icon);
```
I reckon this is the culprit for the `?` issue.

File: totalRP3/Modules/Register/Characters/RegisterCharacteristics.lua [hunk 3]
```diff
@@ -1156,36 +1112,36 @@ function setEditDisplay()
 			end);
 		end);
 
-		_G[frame:GetName() .. "Icon"]:SetScript("onMouseDown", function(self, button)
+		frame.Icon:SetScript("onMouseDown", function(self, button)
```
Comment (Solanya) [inline totalRP3/Modules/Register/Characters/RegisterCharacteristics.lua:1115]:
```suggestion
		frame.Icon:SetScript("OnMouseDown", function(self, button)
```
Consistency

File: totalRP3/Modules/Register/Characters/RegisterUICharacteristics.xml [hunk 4]
```diff
@@ -365,28 +375,92 @@ https://raw.githubusercontent.com/Meorawr/wow-ui-schema/main/UI.xsd">
...
-			<Frame parentKey="Name" inherits="TRP3_TruncatedTextTemplate">
-				<Size x="135"/>
-				<KeyValues>
-					<KeyValue key="fontStringTemplate" value="GameFontNormal"/>
-					<KeyValue key="fontStringJustifyH" value="LEFT"/>
-				</KeyValues>
+		</Frames>
+		<Layers>
+			<Layer level="OVERLAY">
+				<FontString parentKey="Title" font="GameFontNormal" inherits="GameFontNormal" mixin="TRP3_TruncatedTextMixin" justifyH="LEFT">
```
Comment (Solanya) [inline totalRP3/Modules/Register/Characters/RegisterUICharacteristics.xml:381]: is it normal that font and inherits are both there with the same font referenced? Seeing this in a few new fontstrings.

Comment (Meorawr) [reply totalRP3/Modules/Register/Characters/RegisterUICharacteristics.xml]: It was a leftover from when I was deciding how best to deal with the fontstring mixin changes.

File: totalRP3/UI/Browsers/ColorBrowser.lua [hunk 5]
```diff
@@ -0,0 +1,460 @@
...
+	if not TRP3_Colors then
+		TRP3_Colors = {};
+	end
+
+	local _, preset = FindPresetByName(name);
+	local colorString = "#" .. color:GenerateHexColorOpaque();
+
+	if preset then
+		preset.CO = colorString;
+
```
Comment (Solanya) [inline totalRP3/UI/Browsers/ColorBrowser.lua:109]: this empty line in the middle of nowhere

File: totalRP3/UI/ColorSwatch.xml [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+						<Anchor point="TOPLEFT" x="1" y="-1"/>
+						<Anchor point="BOTTOMRIGHT" x="-1" y="1"/>
+					</Anchors>
+					<Color color="BLACK_FONT_COLOR"/>
+				</Texture>
+			</Layer>
+			<Layer level="BACKGROUND">
+				<Texture parentKey="Color" texelSnappingBias="0.0" snapToPixelGrid="false">
+					<Anchors>
+						<Anchor point="TOPLEFT" x="1" y="-1"/>
```
Comment (Solanya) [inline totalRP3/UI/ColorSwatch.xml:33]: I'm guessing it's not really important, but this one doesn't match what's in the OnShow

**Review by Solanya — APPROVED** [review]

**Pull request merged**